0000-event-system.md

```diff
@@ -0,0 +1,109 @@
+- Start Date: 2018-01-16
+- RFC PR: (leave this empty)
+- React Navigation Issue: (leave this empty)
+
+# Summary
+
+Add an event emitter to the navigation prop that emits focus/blur events for the route.
+
+This proposal has been pioneered by @satya164: https://github.com/react-navigation/react-navigation/issues/1363 https://github.com/react-navigation/react-navigation/issues/51
+
+# Basic example
+
+In this example, we focus the first text input when the screen has completed transitioning:
+
+```
+componentDidMount() {
+	this._sub = this.props.navigation.events.addListener(
+		'DidFocus',
+		this._focusFirstTextInput
+	);
+}
+componentWillUnmount() {
+	this._sub.remove();
+}
+```
+
+
+# Motivation
+
+Screens need to know about navigation changes while they are mounted. Consider the following use cases:
+
+### Data and Media subscription
+Screens must know when to request live data or resource-intensive APIs. Media should be able to auto-pause at correct times.
+
+### Unmounting
+Tabs and cards may need to unmount and re-mount contents when they are focused.
+
+### Back button handling
+Screens may want to prevent back actions based on internal state
+
+### Keyboard dismissal
+KB must dismiss when scene *starts* to transition back (mayBlur)
```

Comment:
Another one that is important here is StatusBar, where if it needs to be light on one screen and dark on another, we want to use a sane default for when to change the status bar automatically but let people control it more with route events if they need to